AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -608,23 +663,41 @@ - (ASLayout *)__measureWithSizeRange:(ASSizeRange)constrainedSize
   ASDisplayNodeAssertTrue(_layout.size.width >= 0.0);
   ASDisplayNodeAssertTrue(_layout.size.height >= 0.0);
 
-  // we generate placeholders at measureWithSizeRange: time so that a node is guaranteed to have a placeholder ready to go
-  // also if a node has no size, it should not have a placeholder
-  if (self.placeholderEnabled && [self _displaysAsynchronously] && _layout.size.width > 0.0 && _layout.size.height > 0.0) {
-    if (!_placeholderImage) {
-      _placeholderImage = [self placeholderImage];
-    }
-
-    if (_placeholderLayer) {
-      [self setupPlaceholderLayerContents];
-    }
+  // we generate placeholders at measureWithSizeRange: time so that a node is guaranteed
+  // to have a placeholder ready to go. Also, if a node has no size it should not have a placeholder
+  if (self.placeholderEnabled && [self _displaysAsynchronously] &&
+      _layout.size.width > 0.0 && _layout.size.height > 0.0) {
+    [self __generatePlaceholder];
   }
 
   return _layout;
 }
 
+- (NSArray<_ASDisplayNodePosition *> *)_filterSublayouts:(NSArray<ASLayout *> *)layouts withIndexes:(NSIndexSet *)indexes
+{
+  NSMutableArray<_ASDisplayNodePosition *> *result = [NSMutableArray array];
+  [indexes enumerateIndexesUsingBlock:^(NSUInteger idx, BOOL * _Nonnull stop) {
+    ASDisplayNode *node = (ASDisplayNode *)layouts[idx].layoutableObject;
+    ASDisplayNodeAssertNotNil(node, @"A flattened layout must consist exclusively of node sublayouts");
+    [result addObject:[_ASDisplayNodePosition positionWithNode:node atIndex:idx]];
+  }];
+  return result;
+}
+
 - (void)calculatedLayoutDidChange
 {
+  // subclass override
+}
+
+- (void)__generatePlaceholder
```

Comment:
Not totally clear what this method does, esp. because it doesn't actually create the placeholder layer...should some of this logic just be put into _setupPlaceholderLayerContents?